### PR TITLE
Add deprecation path for renamed training type plugins

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of  #10549

Adds a deprecation path for the training type plugins that we renamed in #11226. 
Example: In 1.6 this usage will be deprecated

```py
from pytorch_lightning.training_type import DDPPlugin
Trainer(strategy=DDPPlugin(find_unused_parameters=False))
```
and should be replaced with

```py
from pytorch_lightning.strategies import DDPStrategy
Trainer(strategy=DDPStrategy(find_unused_parameters=False))
```

TODO:
- add tests for the deprecated classes
- add deprecation warning to the docstring

### Does your PR introduce any breaking changes? If yes, please list them.

No, but after v1.8, when the old plugin aliases are removed, all imports will break. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @tchaton @rohitgr7 @borda

### CODE REVIEW

File: tests/deprecated_api/test_remove_1-8.py [hunk 1]
```diff
@@ -256,3 +269,34 @@ def test_v1_8_0_deprecated_training_type_plugin_property():
...
+        DataParallelPlugin,
+        DDPFullyShardedPlugin,
+        pytest.param(IPUPlugin, marks=RunIf(ipu=True)),
+        DDPShardedPlugin,
+        DDPSpawnShardedPlugin,
+        TPUSpawnPlugin,
+    ],
+)
+def test_v1_8_0_deprecated_training_type_plugin_classes(cls):
+    with pytest.deprecated_call(match="in v1.6 and will be removed in v1.8"):
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:290]: Can you check the name is present?

You can access `cls.__name__` and replace "Plugin" -> "Strategy"

File: tests/deprecated_api/test_remove_1-8.py [hunk 2]
```diff
@@ -256,3 +269,34 @@ def test_v1_8_0_deprecated_training_type_plugin_property():
...
+@pytest.mark.parametrize(
+    "cls",
+    [
+        DDPPlugin,
+        DDP2Plugin,
+        DDPSpawnPlugin,
+        pytest.param(DeepSpeedPlugin, marks=RunIf(deepspeed=True)),
+        DataParallelPlugin,
+        DDPFullyShardedPlugin,
+        pytest.param(IPUPlugin, marks=RunIf(ipu=True)),
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:283]: This won't run at the moment because we only run ipu tests on the ipu test file

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -60,6 +47,21 @@
...
+    DDPSpawnShardedStrategy,
+    DDPSpawnStrategy,
+    DDPStrategy,
+    DeepSpeedStrategy,
+    HorovodStrategy,
+    SingleDeviceStrategy,
+    SingleTPUStrategy,
+    TPUSpawnStrategy,
+)
+from pytorch_lightning.strategies.ipu import IPUStrategy
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:64]: n00b question: how come this is imported separately?

**Review by ananthsub — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC
+
+from pytorch_lightning.strategies import Strategy
+from pytorch_lightning.utilities import rank_zero_deprecation
+
+
+class TrainingTypePlugin(Strategy, ABC):
```
Comment (four4fish) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:20]: Just curious, do we need ABC here?

**Review by four4fish — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-8.py (hunk 2 above)
Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-8.py:283]: Do you think #11170 will land? I can drop the test case completely, move it to the ipu test file, or keep as is

File: tests/deprecated_api/test_remove_1-8.py (hunk 1 above)
Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-8.py]: done

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:20]: This is just to make the typing and linter system happy, doesn't hurt
image

**Review by akihironitta — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/__init__.py [hunk 5]
```diff
@@ -19,21 +19,22 @@
...
-from pytorch_lightning.strategies.horovod import HorovodStrategy
-from pytorch_lightning.strategies.ipu import IPUStrategy
-from pytorch_lightning.strategies.parallel import ParallelStrategy
-from pytorch_lightning.strategies.sharded import DDPShardedStrategy
-from pytorch_lightning.strategies.sharded_spawn import DDPSpawnShardedStrategy
-from pytorch_lightning.strategies.single_device import SingleDeviceStrategy
-from pytorch_lightning.strategies.single_tpu import SingleTPUStrategy
-from pytorch_lightning.strategies.strategy import Strategy
-from pytorch_lightning.strategies.tpu_spawn import TPUSpawnStrategy
+from pytorch_lightning.plugins.training_type.ddp import DDPPlugin
```
Comment (Borda) [inline pytorch_lightning/plugins/__init__.py:22]: does the `DDPPlugin` contain a deprecation message? if not shall we try some mapping it is just about renaming classes:
```
from pytorch_lightning.strategies.ddp import DDPStrategy as DDPPlugin
```

File: pytorch_lightning/plugins/training_type/__init__.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+from pytorch_lightning.plugins.training_type.ddp import DDPPlugin  # noqa: F401
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/__init__.py:1]: then this noqa wont be needed

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/plugins/__init__.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/__init__.py:22]: yes it already contains a deprecation message.

File: tests/deprecated_api/test_remove_1-8.py (hunk 2 above)
Comment (carmocca) [reply tests/deprecated_api/test_remove_1-8.py:283]: I don't know. I would move the test to `test_ipu.py` and add a comment here referencing it.

**Review by tchaton — APPROVED** [review]
LGTM !
